Update internal code for torch.geqrf

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56257 Fix MAGMA qr for empty batched inputs
* #56256 Add cuSOLVER path for torch.linalg.qr
* #56255 Remove size arguments for internal orgqr and geqrf calls
* #56254 Add non-allocating helper function for torch.linalg.qr
* #56253 Add cuBLAS path for batched torch.geqrf
* #56252 Add cuSOLVER path for torch.geqrf
* #56251 Port CUDA torch.geqrf to ATen
* **#56250 Update internal code for torch.geqrf**
* #56249 Port CPU torch.geqrf to ATen

Moved `apply_geqrf` to `BatchLinearAlgebraKernel.cpp`. Added
`geqrf_stub` dispatch.

Differential Revision: [D27907362](https://our.internmc.facebook.com/intern/diff/D27907362)

IvanYashchuk committed:
Update internal code for torch.geqrf

Moved `apply_geqrf` to `BatchLinearAlgebraKernel.cpp`. Added
`geqrf_stub` dispatch.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Update internal code for torch.geqrf"


Moved `apply_geqrf` to `BatchLinearAlgebraKernel.cpp`. Added
`geqrf_stub` dispatch.

[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:
```diff
@@ -330,6 +330,75 @@ void linalg_eigh_kernel(Tensor& eigenvalues, Tensor& eigenvectors, Tensor& infos
 }
 
+/*
+  The geqrf function computes QR decomposition of matrices stored in `input`.
```
"computes the QR"

mruberry commented on aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:
```diff
@@ -330,6 +330,75 @@ void linalg_eigh_kernel(Tensor& eigenvalues, Tensor& eigenvectors, Tensor& infos
 
+/*
+  The geqrf function computes QR decomposition of matrices stored in `input`.
+  However, rather than producing a Q matrix directly, it produces a sequence of
```
This is a really nice description of the operation

IvanYashchuk committed:
Update on "Update internal code for torch.geqrf"


Moved `apply_geqrf` to `BatchLinearAlgebraKernel.cpp`. Added
`geqrf_stub` dispatch.

[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:
```diff
@@ -330,6 +330,75 @@ void linalg_eigh_kernel(Tensor& eigenvalues, Tensor& eigenvectors, Tensor& infos
+template <typename scalar_t>
+static void apply_geqrf(const Tensor& input, const Tensor& tau, int64_t m, int64_t n) {
+#ifndef USE_LAPACK
+  AT_ERROR("geqrf: LAPACK library not found in compilation");
```
TORCH_CHECK and double check this error message for consistency with other LAPACK not found errors, I think we often say something diferent?

mruberry reviewed (approved):
Just two very small nits on this one; nice work!

IvanYashchuk committed:
Update on "Update internal code for torch.geqrf"


Moved `apply_geqrf` to `BatchLinearAlgebraKernel.cpp`. Added
`geqrf_stub` dispatch.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Update internal code for torch.geqrf"


Moved `apply_geqrf` to `BatchLinearAlgebraKernel.cpp`. Added
`geqrf_stub` dispatch.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Update internal code for torch.geqrf"


Moved `apply_geqrf` to `BatchLinearAlgebraKernel.cpp`. Added
`geqrf_stub` dispatch.

[ghstack-poisoned]

IvanYashchuk commented:
@mruberry I updated this PR and https://github.com/pytorch/pytorch/pull/56249, I think they both can be imported now.

IvanYashchuk committed:
Update on "Update internal code for torch.geqrf"


Moved `apply_geqrf` to `BatchLinearAlgebraKernel.cpp`. Added
`geqrf_stub` dispatch.

[ghstack-poisoned]
